edgelet/edgelet-kube/src/runtime.rs

```diff
@@ -194,12 +292,63 @@ where
-        Box::new(
-            spec_to_deployment(&module)
-                .map_err(Error::from)
-                .map(|_| ())
-                .into_future(),
-        )
+        let f = spec_to_deployment(self, &module)
+            .map_err(Error::from)
+            .map(|(name, new_deployment)| {
+                let client_copy = self.client.clone();
+                let namespace_copy = self.namespace().to_owned();
+                let fut = self
+                    .client
+                    .lock()
+                    .expect("Unexpected lock error")
+                    .borrow_mut()
+                    .list_deployments(
+                        self.namespace(),
+                        Some(&name),
+                        Some(&self.device_hub_selector),
+                    )
+                    .map_err(Error::from)
+                    .and_then(move |deployments| {
+                        if let Some(current_deployment) =
+                            deployments.items.into_iter().find(|deployment| {
+                                deployment.metadata.as_ref().map_or(false, |meta| {
+                                    meta.name.as_ref().map_or(false, |n| *n == name)
+                                })
+                            })
+                        {
+                            // found deployment, if the deployment found doesn't match, replace it.
+                            if current_deployment == new_deployment {
+                                Either::A(Either::A(future::ok(())))
+                            } else {
+                                let fut = client_copy
+                                    .lock()
+                                    .expect("Unexpected lock error")
+                                    .borrow_mut()
+                                    .replace_deployment(
+                                        name.as_str(),
+                                        namespace_copy.as_str(),
+                                        &new_deployment,
+                                    )
+                                    .map_err(Error::from)
+                                    .map(|_| ());
+                                Either::A(Either::B(fut))
+                            }
+                        } else {
+                            // Not found - create it.
+                            let fut = client_copy
+                                .lock()
+                                .expect("Unexpected lock error")
+                                .borrow_mut()
+                                .create_deployment(namespace_copy.as_str(), &new_deployment)
+                                .map_err(Error::from)
+                                .map(|_| ());
+                            Either::B(fut)
+                        }
+                    });
+                Either::A(fut)
+            })
+            .unwrap_or_else(|err| Either::B(future::err(err)));
```

Comment:
As before, we should be able to use `into_future` and `flatten` and avoid an `Either`.